### ISSUE TITLE
As a Dataverse Installation Administrator, I want to set up the Provenance system to run alongside Dataverse so that researchers using my installation can track and view provenance

### ISSUE THREAD

**Issue by djbrooke** [IQSS/dataverse#4364]
We'll need to figure out how to integrate and set up the provenance application to run alongside Dataverse. This issue is to cover: 

- determining the best way to integrate the two systems
- security issues
- whether to include prov in the core code or load from somewhere else (like Solr)
- documentation for installation administrators

**Comment by jacksonokuhn** [comment]
https://github.com/End-to-end-provenance/core-provenance-library

**Comment by pdurbin** [comment]
@jacksonokuhn and I have been chatting and I just pushed 9d68fc2 where I added the CPL jar to the project and stubbed out some tests and dev guide updates.

I'm somewhat concerned about all the extra steps required for developers (contrary to our "quick and easy" dev setup desire at #4419) but I guess we'll see if there is any streamlining we can do. I'm only scratching the surface on Mac. I haven't looked a Linux at all.

**Comment by pdurbin** [comment]
Nevermind about 9d68fc2 (native library called by Java). We'll keep that in our back pocket but as discussed at standup we're hoping to run the Prov system as a REST service like Solr. I just pushed 3f34669 where I documented the steps to get this stood up on Ubuntu. Lots of hacks in there and we'll want to switch it to RHEL/CentOS which is our recommended platform so we don't lose the whole "kick the tires with everything installed on a single server" story we have right now. (I opened https://github.com/ProvTools/prov-cpl/issues/4 about how the c++ compiler on stock CentOS 7 is too old.) We'll also want to put at least parts of this process into Jenkins for reproducible building and distributing of the binaries. Lots to do. Nice meeting with @jacksonokuhn and @pameyer this morning. Attached is the Vagrantfile I've been using to spin up an Ubuntu VM: [Vagrantfile.txt](https://github.com/IQSS/dataverse/files/1679540/Vagrantfile.txt)

**Comment by jacksonokuhn** [comment]
The REST API isn't throwing errors anymore, and there are (somewhat skeletal) docs. Will move on to actually returning correct status codes next.

**Comment by pdurbin** [comment]
I'm ready for another developer to walk through the steps I documented in the Dev Guide for how to run the scripts I wrote to get the Prov system stood up in Vagrant. Feedback welcome! Please see https://github.com/IQSS/dataverse/compare/4364-install-prov-flask

**Comment by pdurbin** [comment]
At standup @landreev and I agreed we'd like to write an RPM spec file to build and RPM for the Prov system to make it easier to install on RHEL/CentOS. Unfortunately, I'm having trouble even getting `make` to work on CentOS 7. I opened this issue to track building the RPM: https://github.com/ProvTools/prov-cpl/issues/6

In other news, @matthew-a-dunlap validated the script I wrote yesterday on Ubuntu (a newer, non-LTS version) and added some tweaks to the docs in fd21c12 . I'm still looking for a developer to try `vagrant up` on Mac.

**Comment by matthew-a-dunlap** [comment]
@pdurbin I tried vagrant up on my machine with prov-cpl being located @ ~/Documents/github/prov-cpl , but had the below error. I also tried to run it all @ /prov-cpl but had permission denied errors

`Matthews-MacBook-Pro-3:prov-cpl matthew$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Box 'ubuntu/xenial64' could not be found. Attempting to find and install...
    default: Box Provider: virtualbox
    default: Box Version: >= 0
==> default: Loading metadata for box 'ubuntu/xenial64'
    default: URL: https://vagrantcloud.com/ubuntu/xenial64
==> default: Adding box 'ubuntu/xenial64' (v20180126.0.0) for provider: virtualbox
    default: Downloading: https://vagrantcloud.com/ubuntu/boxes/xenial64/versions/20180126.0.0/providers/virtualbox.box
==> default: Successfully added box 'ubuntu/xenial64' (v20180126.0.0) for 'virtualbox'!
==> default: Importing base box 'ubuntu/xenial64'...
==> default: Matching MAC address for NAT networking...
==> default: Checking if box 'ubuntu/xenial64' is up to date...
==> default: Setting the name of the VM: prov-cpl_default_1517524127042_86970
Vagrant is currently configured to create VirtualBox synced folders with
the `SharedFoldersEnableSymlinksCreate` option enabled. If the Vagrant
guest is not trusted, you may want to disable this option. For more
information on this option, please refer to the VirtualBox manual:

  https://www.virtualbox.org/manual/ch04.html#sharedfolders

This option can be disabled globally with an environment variable:

  VAGRANT_DISABLE_VBOXSYMLINKCREATE=1

or on a per folder basis within the Vagrantfile:

  config.vm.synced_folder '/host/path', '/guest/path', SharedFoldersEnableSymlinksCreate: false
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 5000 (guest) => 7777 (host) (adapter 1)
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Running 'pre-boot' VM customizations...
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: Warning: Connection reset. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Connection reset. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Connection reset. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Connection reset. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Connection reset. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Connection reset. Retrying...
    default:
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default:
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if it's present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
    default: The guest additions on this VM do not match the installed version of
    default: VirtualBox! In most cases this is fine, but in rare cases it can
    default: prevent things such as shared folders from working properly. If you see
    default: shared folder errors, please make sure the guest additions within the
    default: virtual machine match the version of VirtualBox you have installed on
    default: your host and reload your VM.
    default:
    default: Guest Additions Version: 5.0.40
    default: VirtualBox Version: 5.1
==> default: Mounting shared folders...
    default: /prov-cpl => /Users/matthew/Documents/github/prov-cpl
==> default: Running provisioner: shell...
    default: Running: /var/folders/mr/8z8plp8902v43lfcdmp7np_w0000gn/T/vagrant-shell20180201-18134-dbds6.sh
    default: /tmp/vagrant-shell: line 7: syntax error near unexpected token `newline'
    default: /tmp/vagrant-shell: line 7: `<!DOCTYPE html>'
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.`

**Comment by pdurbin** [comment]
@landreev at standup I keep mentioning the RPM spec file I'm hacking on. I just pushed 2f131b6 if you'd like to see the direction I'm taking. Feedback is certainly welcome!

**Comment by matthew-a-dunlap** [comment]
I am able to run vagrant on my mac and connect to it via my ubuntu vm. Thanks @pdurbin for help with vagrant.

I was not able to get CPL installed natively on ubuntu, my last error being: 
```
su postgres -s /bin/sh -c "python cpl-rest.py"
Traceback (most recent call last):
 File "cpl-rest.py", line 2, in <module>
   from CPL import cpl_bundle, cpl_object, cpl_relation, CPLException
ImportError: cannot import name CPLException
```
I am unblocked so resolving this is unnecessary at this point, we will circle back on this at a later date.

**Comment by pdurbin** [comment]
I'm currently blocked on getting the Python bindings to build on CentOS. I pinged @jacksonokuhn about this and opened https://github.com/ProvTools/prov-cpl/issues/7 and pushed my attempt to 6b930af

**Comment by pdurbin** [comment]
Phew! I stuck the Python bindings and more into a single RPM. I have the prov service running on CentOS 7 in Vagrant (the API returns the version, at least) so I made pull request #4461 and I'm moving this issue to code review at https://waffle.io/IQSS/dataverse

**Comment by pdurbin** [comment]
@kcondon @landreev @djbrooke heads up that the RPM I built is only compatible with RHEL/CentOS 7 (el7) because that's the version I built the RPM on because it's my understanding that we're moving to el7 as part of our move to AWS (#4321). I bring this up because I'm aware that most of our QA environments are running el6. Attempting to install the RPM on el6 is expected to fail with an error like the one below. This RPM should be tested on el7 servers. We could also talk about how we build this RPM. Currently, I'm building it in an el7 Vagrant environment. The HMDC Jenkins server runs el6 so even if we can get a job stood up to build the RPM (we'd need to manually `make` and install SWIG 3 on the Jenkins server), I'm concerned that an RPM built on el6 will fail to install to el7. In short, there seem to be ABI (Application Binary Interface) differences between el6 and el7, as shown below.

```
Resolving Dependencies
--> Running transaction check
---> Package dataverse-provenance.x86_64 0:0.1-1 will be installed
--> Processing Dependency: python(abi) = 2.7 for package: dataverse-provenance-0.1-1.x86_64
--> Processing Dependency: libc.so.6(GLIBC_2.14)(64bit) for package: dataverse-provenance-0.1-1.x86_64
--> Processing Dependency: libstdc++.so.6(GLIBCXX_3.4.14)(64bit) for package: dataverse-provenance-0.1-1.x86_64
--> Processing Dependency: libstdc++.so.6(GLIBCXX_3.4.15)(64bit) for package: dataverse-provenance-0.1-1.x86_64
--> Processing Dependency: libodbc.so.2()(64bit) for package: dataverse-provenance-0.1-1.x86_64
--> Running transaction check
---> Package dataverse-provenance.x86_64 0:0.1-1 will be installed
--> Processing Dependency: python(abi) = 2.7 for package: dataverse-provenance-0.1-1.x86_64
--> Processing Dependency: libc.so.6(GLIBC_2.14)(64bit) for package: dataverse-provenance-0.1-1.x86_64
--> Processing Dependency: libstdc++.so.6(GLIBCXX_3.4.14)(64bit) for package: dataverse-provenance-0.1-1.x86_64
--> Processing Dependency: libstdc++.so.6(GLIBCXX_3.4.15)(64bit) for package: dataverse-provenance-0.1-1.x86_64
---> Package unixODBC.x86_64 0:2.2.14-14.el6 will be installed
--> Finished Dependency Resolution
Error: Package: dataverse-provenance-0.1-1.x86_64 (/dataverse-provenance-0.1-1.x86_64)
           Requires: libstdc++.so.6(GLIBCXX_3.4.15)(64bit)
Error: Package: dataverse-provenance-0.1-1.x86_64 (/dataverse-provenance-0.1-1.x86_64)
           Requires: libc.so.6(GLIBC_2.14)(64bit)
Error: Package: dataverse-provenance-0.1-1.x86_64 (/dataverse-provenance-0.1-1.x86_64)
           Requires: python(abi) = 2.7
           Installed: python-2.6.6-66.el6_8.x86_64 (@updates)
               python(abi) = 2.6
           Available: python34-3.4.5-2.el6.i686 (epel)
               python(abi) = 3.4
           Available: python34-3.4.5-4.el6.i686 (epel)
               python(abi) = 3.4
Error: Package: dataverse-provenance-0.1-1.x86_64 (/dataverse-provenance-0.1-1.x86_64)
           Requires: libstdc++.so.6(GLIBCXX_3.4.14)(64bit)
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
[root@dvnweb-vm2 ~]# 
[root@dvnweb-vm2 ~]# cat /etc/redhat-release 
CentOS release 6.9 (Final)
[root@dvnweb-vm2 ~]# 
```

**Comment by pdurbin** [comment]
@jacksonokuhn and I discussed this issue after standup. He's able to build the RPM, which is great. Once his repo is ready, we can close pull request #4461 and write new instructions (a new pull request) linking to the RPM hosted in the prov-cpl repo.

**Comment by djbrooke** [comment]
Moving the "backlog" column just for organizational purposes.

**Comment by djbrooke** [comment]
Will close for now, will reopen if we decide to take on a similar approach with further provenance work.
